[OptBisect][ADT] Add support for running ranges of passes and introduce IntegerInclusiveInterval ADT

YonahGoldberg opened the pull request:
This PR does two things:
1. Introduce `-opt-bisect=<inclusive_integer_interval>` for running ranges of passes, where `inclusive_integer_interval` could be:
- `-1` to run all pases
- `0` to run no passes
- `10-12` to run passes 10 through 12
- `10-12,15-18 `to run passes 10 through 10 and 15 through 18 

`-opt-bisect-limit=N` now maps to `-opt-bisect=1-N`

2. Introduces `IntegerInclusiveInterval` in ADT to represent these intervals and the ability to parse string ranges (from the command line). The motivation to move this into ADT is that both `opt-bisect` and `debug-counter` need to parse intervals from the command line. `reduce-chunk-list` also relied on a interval object, so I refactored that to use `IntegerInclusiveInterval` as well.

YonahGoldberg committed:
opt bisect skip

YonahGoldberg committed:
format

YonahGoldberg committed:
comment

YonahGoldberg commented:
@AlexMaclean @nikic

YonahGoldberg committed:
newline

AlexMaclean reviewed (approved):
Nice, I personally use this feature in our down-stream fork and find it to be a nice complement to --opt-bisect-limit when investigating correctness issues. Please wait for some more review prior to landing.

YonahGoldberg committed:
renaming

YonahGoldberg commented on llvm/test/Other/opt-bisect-skip.ll:
```diff
@@ -0,0 +1,15 @@
+; Test that verifies functionality for -opt-bisect-skip
+
+; RUN: opt -O1 -opt-bisect-skip=3 -opt-bisect-skip=7 %s 2>&1 | FileCheck %s --check-prefix=CHECK-DISABLE-PASS
```
Fixed that.

YonahGoldberg commented on llvm/lib/IR/OptBisect.cpp:
```diff
@@ -37,6 +37,11 @@ static cl::opt<int> OptBisectLimit("opt-bisect-limit", cl::Hidden,
                                    cl::desc("Maximum optimization to perform"));
 
+static cl::opt<int> OptBisectSkip(
+    "opt-bisect-skip", cl::Hidden, cl::init(OptBisect::Disabled), cl::Optional,
```
I changed to your naming scheme bc I agree it's better, open to other suggestions though.

YonahGoldberg committed:
format

nikic reviewed (commented):
I think it would be better to make opt-bisect work the same way that debug-counter works nowadays, that is by specifying a list of ranges. So that you can use something like `-opt-bisect=1-1000` to get something similar to the current opt-bisect-limit, but also `-opt-bisect=1-10,20-30` to skip a certain range of passes, etc.

YonahGoldberg committed:
range support

YonahGoldberg committed:
small changes

YonahGoldberg committed:
remove opt-disable-indices

YonahGoldberg committed:
remove

YonahGoldberg committed:
cleanup

YonahGoldberg committed:
format

YonahGoldberg committed:
use Expect for Range

YonahGoldberg committed:
format

YonahGoldberg committed:
fix hang

YonahGoldberg committed:
fix error handling tests

YonahGoldberg committed:
format

YonahGoldberg committed:
cleanup

YonahGoldberg committed:
clang format

YonahGoldberg committed:
fix minor formatting bug

YonahGoldberg committed:
small bug

YonahGoldberg committed:
format

YonahGoldberg committed:
allow -opt-bisect=0 and -opt-bisect=-1

YonahGoldberg committed:
comment change

YonahGoldberg committed:
format

YonahGoldberg committed:
cleanup

YonahGoldberg committed:
format

YonahGoldberg committed:
more cleanup

YonahGoldberg committed:
format

YonahGoldberg committed:
range test cleanup

YonahGoldberg committed:
format

YonahGoldberg committed:
move range from struct -> class

YonahGoldberg committed:
better range size function

YonahGoldberg committed:
range size checks

YonahGoldberg committed:
format

YonahGoldberg committed:
format

YonahGoldberg commented:
Hey @nikic I think I addressed all your comments, thanks for the feedback :)

The most major change is I decided to make Range into a class to preserve the Begin <= End invariant.

YonahGoldberg commented:
@nikic @Artem-B gentle ping for review.

nikic commented on llvm/lib/Support/Range.cpp:
```diff
@@ -0,0 +1,117 @@
+  else {
+    bool IsFirst = true;
+    for (const Range &R : Ranges) {
+      if (!IsFirst)
```
Can use ListSeparator for this.

nikic reviewed (commented):
This looks reasonable to me, but an ADT reviewer will probably want to take a look.

I have one high level question: Why is the Range here signed? It seems like in this context negative numbers are not relevant. Is this just because you expect it to be useful for other use cases? (Worth noting that for "real" ranges we have ConstantRange and ConstantRangeList in IR.)

YonahGoldberg commented on llvm/include/llvm/Support/Range.h:
```diff
@@ -0,0 +1,126 @@
+  }
+
+  /// Get the size of this range.
+  uint64_t size() const {
```
No, I can remove it.

YonahGoldberg commented:
> This looks reasonable to me, but an ADT reviewer will probably want to take a look.
> 
> I have one high level question: Why is the Range here signed? It seems like in this context negative numbers are not relevant. Is this just because you expect it to be useful for other use cases? (Worth noting that for "real" ranges we have ConstantRange and ConstantRangeList in IR.)

It's signed because I expect it might be useful in other cases, yes.

Do you think it would make more sense to just extend ConstantRangeList? I could create a RangeList::fromRangeString function. I don't actually need the range bounds to be mutable so I think this actually might make the most sense.

CC: @arsenm @haopliu

YonahGoldberg committed:
use list separator

YonahGoldberg committed:
remove size

YonahGoldberg committed:
redundant include

YonahGoldberg committed:
IntegerInclusiveInterval

YonahGoldberg committed:
format

YonahGoldberg committed:
more pruning range

YonahGoldberg committed:
format

YonahGoldberg committed:
format

YonahGoldberg committed:
fix tests

YonahGoldberg committed:
renaming

YonahGoldberg committed:
format

YonahGoldberg committed:
nits

YonahGoldberg committed:
format

YonahGoldberg committed:
forgot to update header

YonahGoldberg committed:
format
